kahypar/partition/bin_packing/bin_packer.h

```diff
@@ -0,0 +1,124 @@
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with KaHyPar.  If not, see <http://www.gnu.org/licenses/>.
+ *
+******************************************************************************/
+
+#pragma once
+
+#include "kahypar/definitions.h"
+#include "kahypar/partition/bin_packing/i_bin_packer.h"
+#include "kahypar/partition/bin_packing/bin_packing_utils.h"
+#include "kahypar/partition/bin_packing/bin_packing_algorithms.h"
+#include "kahypar/partition/bin_packing/prepacking.h"
+
+namespace kahypar {
+namespace bin_packing {
+// Implementation of IBinPacker for a generic bin packing algorithm.
+template< class BPAlgorithm = WorstFit >
+class BinPacker final : public IBinPacker {
+ public:
+  BinPacker(Hypergraph& hypergraph, Context& context) : _hypergraph(hypergraph), _context(context) {}
+
+ private:
+  void prepackingImpl(const BalancingLevel level) override {
+    ASSERT(!_hypergraph.containsFixedVertices(), "No fixed vertices allowed before prepacking.");
+    ASSERT(level != BalancingLevel::STOP, "Invalid balancing level: STOP");
+
+    const PartitionID rb_range_k = _context.partition.rb_upper_k - _context.partition.rb_lower_k + 1;
+    const HypernodeWeight max_bin_weight = floor(_context.initial_partitioning.current_max_bin_weight * (1.0 + _context.initial_partitioning.bin_epsilon));
+
+    if (level == BalancingLevel::heuristic) {
+      bin_packing::calculateHeuristicPrepacking<BPAlgorithm>(_hypergraph, _context, rb_range_k, max_bin_weight);
+    } else if (level == BalancingLevel::guaranteed) {
+      for (size_t i = 0; i < static_cast<size_t>(_context.initial_partitioning.k); ++i) {
+        const HypernodeWeight lower = _context.initial_partitioning.perfect_balance_partition_weight[i];
+        const HypernodeWeight upper = _context.initial_partitioning.num_bins_per_part[i] * max_bin_weight;
+        HypernodeWeight& border = _context.initial_partitioning.upper_allowed_partition_weight[i];
+
+        // The performance of the prepacking algorithm depends on the perfect part weight, the allowed weight for each bin
+        // and the allowed part weight, which must be between those, i.e. lower <= border <= upper. If border has a value close
+        // to lower, the possibilities of the initial partitioning algorithm are rather restricted, but only few vertices need
+        // to be prepacked. On the other hand, if border has a value close to upper, almost every vertex will be prepacked,
+        // resulting in a bisection with a poor cut.
+        // To avoid this, we check for this case and adjust the allowed part weight if necessary.
+        if (upper - border < (border - lower) / 10) {
+          border = (lower + upper) / 2;
+          _context.partition.epsilon = static_cast<double>(border) / static_cast<double>(lower) - 1.0;
+        }
+      }
+
+      bin_packing::calculateExactPrepacking<BPAlgorithm>(_hypergraph, _context, rb_range_k, max_bin_weight);
```

Comment:
`bin_packing::` shouldn't be necessary here, since you are already in the `bin_packing` namespace.